pkg/ccl/backupccl/targets.go

```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
 	return fullClusterDescs, fullClusterDBIDs, nil
 }
 
+// checkMissingIntroducedSpans asserts that each backup's IntroducedSpans
+// contain all tables that require an introduction (i.e. a backup from ts=0), if
+// they are also restore targets. This specifically entails checking the
+// following invariant on each table we seek to restore: if the nth backup
+// contains a table in its backupManifest.Descriptors field but the n-1th does
+// not, then that table must be covered by the nth backup's
+// manifest.IntroducedSpans field. This check only applies to backups with a
+// start time that is less than the restore AOST.
+func checkMissingIntroducedSpans(
+	restoringDescs []catalog.Descriptor,
+	mainBackupManifests []BackupManifest,
+	endTime hlc.Timestamp,
+	codec keys.SQLCodec,
+) error {
+	// Gather the tables we'd like to restore.
+	requiredTables := make(map[descpb.ID]struct{})
+	for _, restoringDesc := range restoringDescs {
+		if _, ok := restoringDesc.(catalog.TableDescriptor); ok {
+			requiredTables[restoringDesc.GetID()] = struct{}{}
+		}
+	}
+
+	for i, b := range mainBackupManifests {
+		if !endTime.IsEmpty() && endTime.Less(b.StartTime) {
+			// No need to check a backup that began after the restore AOST.
+			return nil
+		}
+
+		if i == 0 {
+			// The full backup does not contain reintroduced spans
+			continue
+		}
+
+		// Gather the tables included in the previous backup
+		prevTables := make(map[descpb.ID]struct{})
+		for _, desc := range mainBackupManifests[i-1].Descriptors {
+			if table, _, _, _ := descpb.FromDescriptor(&desc); table != nil {
+				prevTables[table.GetID()] = struct{}{}
+			}
+		}
+
+		// Gather the introduced tables
+		tablesIntroduced := make(map[descpb.ID]struct{})
+		for _, span := range mainBackupManifests[i].IntroducedSpans {
+			_, tableID, err := codec.DecodeTablePrefix(span.Key)
+			if err != nil {
+				return err
+			}
+			tablesIntroduced[descpb.ID(tableID)] = struct{}{}
+		}
+
+		// If the table in the current backup was not in prevBackup.Descriptors,
+		// then it needs to be introduced.
+		for _, desc := range mainBackupManifests[i].Descriptors {
+			if table, _, _, _ := descpb.FromDescriptor(&desc); table != nil {
+
+				if _, required := requiredTables[table.GetID()]; !required {
+					continue
+				}
+
+				_, inPrevBackup := prevTables[table.GetID()]
+
+				_, introduced := tablesIntroduced[table.GetID()]
+
+				if !inPrevBackup && !introduced {
+					tableError := errors.Newf("table %v was not properly backed up after its import"+
+						" completed. To continue the restore, "+
+						"either 1) restore from a different full backup; 2) restore to a system time before"+
```

Comment:
Should we say something like:

"table %v was not properly backed up because of <issue_num>. To restore this table one can:

1) Restore from the full backup.
2) Restore AOST <= %s, (I think we can get the actual timestamp from the prevBackup manifest that applies to this table?). Maybe we can stash it in `inPrevBackup` as the value. IIUC that is the last valid backup for this table.

Personally, I'd drop 3) but maybe @dt can chime in here.

Reply:
I like this: 

> table %v was not properly backed up because of <issue_num>.

I don't quite understand your change to suggestion 1. My original intention in suggestion 1 is that any other backup chain in the collection will not have this error.

For 2: The latest possible timestamp is right before the import ended, which is info we cannot easily retrieve in this function. I think it would require iterating over manifest.DescriptorChanges, finding the descriptor change that corresponds to turning the table back online, and returning descChange.ModificationTime - someDelta to the user. I could also return the end time of the last incremental backup that observed the table offline (much easier).

For 3: I could see a use case for 3. suppose the user doesn't actually need the corrupt table in the database restore, but really wants to restore their data to that AOST.

Reply:
> 1. My original intention in suggestion 1 is that any other backup chain in the collection will not have this error.

Even the full backup and some incrementals of this errant chain are restoreable iiuc? Maybe we don't need point 1 if we make 2 clear enough. For point 2 isn't the last valid backup for this table the incremental before the incremental in which this table appeared online but we forgot to reintroduce it? In that case can we just take the last incrementals EndTime and say any ts before this is valid.

Reply:
resolved offline.